File: Java/PDP/src/producerconsumer/ProducerConsumerExample.java
package producerconsumer;

import java.util.ArrayList;
import java.util.List;

public class ProducerConsumerExample {
	public static void main(String[] args) {
		
		int numberProducer = 10;
		int numberConsumer = 10;		
		int numberItens = 10;
		
		Buffer<Integer> buffer = new Buffer<Integer>();
		
		List<Producer> lstProducer = new ArrayList<Producer>(numberProducer);
		List<Consumer> lstConsumer = new ArrayList<Consumer>(numberConsumer);
		
		for(int i = 0; i < numberProducer; i++) {
			lstProducer.add(new Producer(i, buffer));
		}
		
		for(int i = 0; i < numberConsumer; i++) {
			lstConsumer.add(new Consumer(i, buffer));
		}
		
		for(Producer p : lstProducer) {
			p.start();
		}
		
		for(Consumer c : lstConsumer) {
			c.start();
		}
		
		try {
			for(Producer p : lstProducer) {
				p.join();
			}
			
			for(Consumer c : lstConsumer) {
				c.join();
			}
		} catch (Exception e) {
			// TODO: handle exception
		}		
		
	}
	
	
}

File: CSharp/Monitor/ProducerConsumerExample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace ProducerConsumer
{
    public class ProducerConsumerExample
    {
        static void Main(string[] args)
        {
            int numberProducer = 10;
		    int numberConsumer = 10;		
		    int numberItens = 10;
		
		    Buffer<int> buffer = new Buffer<int>();
		
            List<Thread> lstProducer = new List<Thread>();
		    List<Thread> lstConsumer = new List<Thread>();
		
		    for(int i = 0; i < numberProducer; i++) {
			    lstProducer.Add(new Thread( new Producer(i, buffer).Run ));
		    }
		
		    for(int i = 0; i < numberConsumer; i++) {
			    lstConsumer.Add(new Thread ( new Consumer(i, buffer).Run ));
		    }
		
		    foreach(Thread p in lstProducer) {
			    p.Start();
		    }
		
		    foreach(Thread c in lstConsumer) {
			    c.Start();
		    }
		
		    try {
			    foreach(Thread p in lstProducer) {
				    p.Join();
			    }
			
			    foreach(Thread c in lstConsumer) {
				    c.Join();
			    }
		    } catch (Exception e) {
			    // TODO: handle exception
		    }
        }
    }
}
